### ISSUE TITLE
Mixing top level windows and dock windows

### ISSUE THREAD

**Issue by dgregorius** [ocornut/imgui#3152]
**Version/Branch of Dear ImGui:**

Version: 17601
Branch: docking

Back-ends: imgui_impl_glfw.cpp +imgui_impl_opengl3.cpp _
Compiler: VS 2019
Operating System: Windows 10

**My Issue/Question:**

I have an application which consists of several top level windows and each top level window consists of several dock windows. I would like the top level windows have an OS title bar and no auto merge when they overlap. The dock widgets on the other hand should have no decoration and can auto merge.

It seems I can have either one or the other. E.g. if I initialize Dear ImGUI like below get the correct top level behavior, but the dock windows are wrong:

	ImGuiIO& IO = ImGui::GetIO();
	IO.ConfigFlags |= ImGuiConfigFlags_DockingEnable;
	IO.ConfigFlags |= ImGuiConfigFlags_ViewportsEnable;
	IO.ConfigViewportsNoAutoMerge = true;     
	IO.ConfigViewportsNoTaskBarIcon = false;    
	IO.ConfigViewportsNoDecoration = false; 
	IO.ConfigViewportsNoDefaultParent = true;

 If I use the defaults, the dock windows are correct, but the top level window are not working as expected

	ImGuiIO& IO = ImGui::GetIO();
	IO.ConfigFlags |= ImGuiConfigFlags_DockingEnable;
	IO.ConfigFlags |= ImGuiConfigFlags_ViewportsEnable;
	IO.ConfigViewportsNoAutoMerge = false;     
	IO.ConfigViewportsNoTaskBarIcon = false;    
	IO.ConfigViewportsNoDecoration = true; 
	IO.ConfigViewportsNoDefaultParent = false;

Is their a way to specify the above flags per window instead of globally? I think this would solve my issue. It would be nice to have this kind of control I think. In particular since the functionality seems to be already there

Another option would be to create several OS windows myself (in my case with GLFW) and then bind my OS windows to a specific viewport. Unfortunately it seems I cannot create viewports for OS windows myself or find them if they would be auto created.

Finally I was thinking to use a context for each top level window, but again I cannot pass the associated window/viewport.

Any suggestion are welcome!

**Screenshots/Video**

The correct top level behavior

![Capture1](https://user-images.githubusercontent.com/7042834/80046676-235b6180-84c0-11ea-8bdc-fda05631fc0a.PNG)

Incorrect docking windows. The 'Console' window is a docking window and should not have the OS title bar.

![Capture2](https://user-images.githubusercontent.com/7042834/80046830-9bc22280-84c0-11ea-852f-23bfb7f12e20.PNG)

**Comment by ocornut** [comment]
> Is their a way to specify the above flags per window instead of globally? I think this would solve my issue.

Yes. You can use SetNextWindowClass() with a class that override viewport flags (clear/set specific flags).

**Comment by dgregorius** [comment]
Thanks, Omar. This works great. Just for reference if anybody is interested in this. I applied these two settings:

1) When you enable viewport/docking also disable merging

	IO.ConfigFlags |= ImGuiConfigFlags_DockingEnable;
	IO.ConfigFlags |= ImGuiConfigFlags_ViewportsEnable;
++	IO.ConfigViewportsNoAutoMerge = true;  
 ++	IO.ConfigViewportsNoTaskBarIcon = true; 

I only want one 

2) When you pass the window class clear the no decoration flag

		ImGuiWindowClass WindowClass;
		WindowClass.ViewportFlagsOverrideClear = ImGuiViewportFlags_NoDecoration | ImGuiViewportFlags_NoTaskBarIcon;
		ImGui::SetNextWindowClass( &WindowClass );

This should give you OS decoration for every top level window and also a task bar icon. If you want just a single task bar icon for your whole application omit the ImGuiViewportFlags_NoTaskBarIcon flag.

HTH,
-Dirk

**Comment by ocornut** [comment]
Great to hear!

FYI I realize that many of the advanced Viewports flags may be a little inconsistent or hard to parse (since you "set" or "clear" a "NoXXXX" flag), I am expecting to eventually have a api breaking pass at shuffling and renaming some flags, but it'll be reasonably easy to fix things when/if that happens.

**Comment by dgregorius** [comment]
Yes, the naming is indeed not the sexiest. :) But you can always look at the source. So there is no real excuse. Either way, I just tested it more and it works flawlessly.
